### PR TITLE
fix buffered download

### LINKED ISSUE #24589: cuda (many versions): OverflowError: signed integer is greater than maximum on Ubuntu 20.04, 21.04
Cannot install, at least,`cuda@10.1.243:11.3.1` due to `OverflowError: signed integer is greater than maximum`:

Using:
* `spack@develop` (a7b6149cc02713de55cb5719b9dab2dfc0476303 from `Tue Jun 29 00:33:35 2021 -0400`)
* Ubuntu 20.04, Ubuntu 21.04
* GCC 9.3.0, GCC 10.3.0
* Python 3.8.5, Python 3.9.5

```
$> spack install cuda@11.3.1
...
==> Installing cuda-11.3.1-5japwam43c6ocqge56qnjgjs6c67spnb
==> Fetching https://cache.e4s.io/build_cache/linux-ubuntu20.04-x86_64/gcc-9.3.0/cuda-11.3.1/linux-ubuntu20.04-x86_64-gcc-9.3.0-cuda-11.3.1-5japwam43c6ocqge56qnjgjs6c67spnb.spack
==> Error: Failed to install cuda due to OverflowError: signed integer is greater than maximum
==> Error: signed integer is greater than maximum
```

```
$> spack -d install cuda@11.3.1
...
==> [2021-06-29-08:14:25.453897] Installing cuda-11.3.1-5japwam43c6ocqge56qnjgjs6c67spnb
==> [2021-06-29-08:14:25.453935] Searching for binary cache of cuda-11.3.1-5japwam43c6ocqge56qnjgjs6c67spnb
==> [2021-06-29-08:14:25.453997] Reading config file /spack/etc/spack/defaults/mirrors.yaml
==> [2021-06-29-08:14:25.586708] Did not find linux-ubuntu20.04-x86_64-gcc-9.3.0-cuda-11.3.1-5japwam43c6ocqge56qnjgjs6c67spnb.spec.yaml on https://mirror.spack.io/build_cache/linux-ubuntu20.04-x86_64-gcc-9.3.0-cuda-11.3.1-5japwam43c6ocqge56qnjgjs6c67spnb.spec.yaml
  Download failed: HTTP Error 404: Not Found
==> [2021-06-29-08:14:25.588309] Creating stage lock build_cache
==> [2021-06-29-08:14:25.588690] Checking existence of https://cache.e4s.io/build_cache/linux-ubuntu20.04-x86_64/gcc-9.3.0/cuda-11.3.1/linux-ubuntu20.04-x86_64-gcc-9.3.0-cuda-11.3.1-5japwam43c6ocqge56qnjgjs6c67spnb.spack
==> [2021-06-29-08:14:25.673795] Fetching https://cache.e4s.io/build_cache/linux-ubuntu20.04-x86_64/gcc-9.3.0/cuda-11.3.1/linux-ubuntu20.04-x86_64-gcc-9.3.0-cuda-11.3.1-5japwam43c6ocqge56qnjgjs6c67spnb.spack
==> [2021-06-29-08:14:25.807447] Flagging cuda-11.3.1-5japwam43c6ocqge56qnjgjs6c67spnb as failed: signed integer is greater than maximum
==> [2021-06-29-08:14:25.810523] Error: Failed to install cuda due to OverflowError: signed integer is greater than maximum
==> [2021-06-29-08:14:25.891387] Skip view update, this environment does not maintain a view
Traceback (most recent call last):
  File "/spack/bin/spack", line 77, in <module>
    sys.exit(spack.main.main())
  File "/spack/lib/spack/spack/main.py", line 771, in main
    return _invoke_command(command, parser, args, unknown)
  File "/spack/lib/spack/spack/main.py", line 499, in _invoke_command
    return_val = command(parser, args)
  File "/spack/lib/spack/spack/cmd/install.py", line 471, in install
    install_specs(args, kwargs, zip(abstract_specs, specs))
  File "/spack/lib/spack/spack/cmd/install.py", line 271, in install_specs
    env.install_specs(specs_to_install, args=cli_args, **kwargs)
  File "/spack/lib/spack/spack/environment.py", line 1608, in install_specs
    builder.install()
  File "/spack/lib/spack/spack/installer.py", line 1574, in install
    self._install_task(task)
  File "/spack/lib/spack/spack/installer.py", line 1127, in _install_task
    _install_from_cache(pkg, cache_only, explicit, unsigned,
  File "/spack/lib/spack/spack/installer.py", line 274, in _install_from_cache
    installed_from_cache = _try_install_from_binary_cache(
  File "/spack/lib/spack/spack/installer.py", line 405, in _try_install_from_binary_cache
    return _process_binary_cache_tarball(pkg, binary_spec, explicit, unsigned,
  File "/spack/lib/spack/spack/installer.py", line 365, in _process_binary_cache_tarball
    tarball = binary_distribution.download_tarball(
  File "/spack/lib/spack/spack/binary_distribution.py", line 1052, in download_tarball
    stage.fetch()
  File "/spack/lib/spack/spack/stage.py", line 490, in fetch
    self.fetcher.fetch()
  File "/spack/lib/spack/spack/fetch_strategy.py", line 73, in wrapper
    return fun(self, *args, **kwargs)
  File "/spack/lib/spack/spack/fetch_strategy.py", line 318, in fetch
    partial_file, save_file = self._fetch_from_url(url)
  File "/spack/lib/spack/spack/fetch_strategy.py", line 356, in _fetch_from_url
    return self._fetch_urllib(url)
  File "/spack/lib/spack/spack/fetch_strategy.py", line 73, in wrapper
    return fun(self, *args, **kwargs)
  File "/spack/lib/spack/spack/fetch_strategy.py", line 385, in _fetch_urllib
    _data = response.read()
  File "/usr/lib/python3.8/http/client.py", line 471, in read
    s = self._safe_read(self.length)
  File "/usr/lib/python3.8/http/client.py", line 612, in _safe_read
    data = self.fp.read(amt)
  File "/usr/lib/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
  File "/usr/lib/python3.8/ssl.py", line 1241, in recv_into
    return self.read(nbytes, buffer)
  File "/usr/lib/python3.8/ssl.py", line 1099, in read
    return self._sslobj.read(len, buffer)
OverflowError: signed integer is greater than maximum
```

@ax3l @Rombur @fspiga

### PR DESCRIPTION
**Pull request by haampie**
Closes https://github.com/spack/spack/issues/24589

For anybody reviewing, the only relevant change is https://github.com/spack/spack/pull/24623/commits/ff3f3633edc6ffb5050da90db00727d827f11cab, the rest is flake8 import order noise

### CODE REVIEW

**Comment by adamjstewart** [comment]
This may be of interest, but this is how PyTorch/torchvision handle large downloads with urlib:
https://github.com/pytorch/vision/blob/master/torchvision/datasets/utils.py#L30,L38

**Comment by haampie** [comment]
shutil has the advantage that we don't have to define the buffer size like they do

**Review by becker33 — DISMISSED** [review]

**Review by loulawrence — DISMISSED** [review]

**Comment by eugeneswalker** [comment]
Pipeline failing at least in part here because of network/infrastructure issues on `uo-godzilla` runner. @scottwittenburg 
* https://gitlab.spack.io/spack/spack/-/jobs/555763
* https://gitlab.spack.io/spack/spack/-/jobs/555758

I've disabled `uo-godzilla` until the network issues are resolved and am re-starting the failed jobs now.

**Comment by eugeneswalker** [comment]
> Pipeline failing at least in part here because of network/infrastructure issues on `uo-godzilla` runner. @scottwittenburg
> 
> * https://gitlab.spack.io/spack/spack/-/jobs/555763
> * https://gitlab.spack.io/spack/spack/-/jobs/555758

I've disabled `uo-godzilla` and am re-starting the failed jobs manually.

**Comment by haampie** [comment]
@alalazo can you hit merge again? the last commit is about whitespace (assuming the issues above are spurious)

File: lib/spack/spack/cmd/ci.py [hunk 1]
```diff
@@ -24,7 +24,6 @@
 import spack.util.url as url_util
 import spack.util.web as web_util
 
-
```
Comment (alalazo) [inline lib/spack/spack/cmd/ci.py]: This change is spurious and can be removed from the PR

File: lib/spack/spack/test/ci.py [hunk 2]
```diff
@@ -3,26 +3,26 @@
...
 import spack.paths as spack_paths
 import spack.spec as spec
 import spack.util.gpg
-
-import spack.ci_optimization as ci_opt
-import spack.ci_needs_workaround as cinw
 import spack.util.spack_yaml as syaml
-import itertools as it
-import collections
+
```
Comment (alalazo) [inline lib/spack/spack/test/ci.py]: I think also changes in this file are spurious and can be removed.

**Review by alalazo — CHANGES_REQUESTED** [review]
Since CI is failing, can you rebase and keep only changes that are relevant to the PR?

**Review by alalazo — APPROVED** [review]

**Pull request merged**